File: front/src/model/events/HarvestingEvent.ts
import { Fingerprint } from "../Crypto";

namespace HarvestingEvent {
  export type Accepted = {
    readonly tag: "accepted";
    readonly member: Fingerprint;
  };

  export type Refused = {
    readonly tag: "refused";
    readonly member: Fingerprint;
  };

  export type Invalid = {
    readonly tag: "invalid";
  };
}

export type HarvestingEvent =
  | HarvestingEvent.Accepted
  | HarvestingEvent.Refused
  | HarvestingEvent.Invalid;

File: back/src/main/scala/models/HarvestingEvent.scala
package chrilves.kuzh.back.models

import io.circe.*
import io.circe.syntax.*

enum HarvestingEvent:
  case Accepted(member: Member.Fingerprint)
  case Invalid

object HarvestingEvent:
  given harvestingEventEncoder: Encoder[HarvestingEvent] with
    final def apply(e: HarvestingEvent): Json =
      e match
        case Accepted(member) =>
          Json.obj(
            "tag"    -> "accepted".asJson,
            "member" -> member.asJson
          )
        case Invalid =>
          Json.obj(
            "tag" -> "invalid".asJson
          )
